### ISSUE TITLE
hddtemp module fails: received data doesn't have needed records

### ISSUE THREAD

**Issue by adrfantini** [firehol/netdata#3683]
The hddtemp module (default config) does not seem to be able to get data from the netdata daemon on my system:
```
$ /usr/lib/netdata/plugins.d/python.d.plugin 1 debug trace hddtemp


2018-05-04 07:52:13: python.d INFO: plugin: main: Using python 3
2018-05-04 07:52:13: python.d DEBUG: plugin: main: module load source: 'hddtemp' => [OK]
2018-05-04 07:52:13: python.d DEBUG: plugin: main: module load config: 'hddtemp' => [OK]
2018-05-04 07:52:13: python.d DEBUG: plugin: main: job initialization: 'hddtemp localhost' => ['OK']
2018-05-04 07:52:13: python.d DEBUG: plugin: main: job initialization: 'hddtemp localipv4' => ['OK']
2018-05-04 07:52:13: python.d DEBUG: plugin: main: job initialization: 'hddtemp localipv6' => ['OK']
2018-05-04 07:52:13: python.d DEBUG: plugin: main: module status: 'hddtemp' => [OK] (jobs: 3)
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: No unix socket specified. Trying TCP/IP socket.
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: No request specified. Using: ""
2018-05-04 07:52:13: python.d INFO: hddtemp: localhost: Autodetecting disks
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: Creating socket to "::1", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: connecting socket to "::1", port 7634
2018-05-04 07:52:13: python.d ERROR: hddtemp: localhost: Failed to connect to "::1", port 7634, error: [Errno 111] Connection refused
2018-05-04 07:52:13: python.d ERROR: hddtemp: localhost: Traceback (most recent call last):
  File "/usr/lib/netdata/python.d/python_modules/bases/FrameworkServices/SocketService.py", line 61, in _connect2socket
    self._sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: closing socket
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: Creating socket to "192.168.1.100", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: connecting socket to "192.168.1.100", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: connected to "192.168.1.100", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: set socket timeout to: 5.0
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: received data
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: server closed the connection
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: closing socket
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localhost: final response: |/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-HUP722020APA330_BFGWU7WF|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-WDC_WD10EARS-00Y5B1_WD-WCAV5R693922|WDC WD10EARS-00Y5B1|SLP|*|
2018-05-04 07:52:13: python.d ERROR: hddtemp: localhost: received data doesn't have needed records
2018-05-04 07:52:13: python.d INFO: hddtemp: localhost: check() => [FAILED]
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: No unix socket specified. Trying TCP/IP socket.
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: No request specified. Using: ""
2018-05-04 07:52:13: python.d INFO: hddtemp: localipv4: Autodetecting disks
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: Creating socket to "127.0.0.1", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: connecting socket to "127.0.0.1", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: connected to "127.0.0.1", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: set socket timeout to: 5.0
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: receiving response
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: received data
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: receiving response
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: server closed the connection
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: closing socket
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: final response: |/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-HUP722020APA330_BFGWU7WF|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-WDC_WD10EARS-00Y5B1_WD-WCAV5R693922|WDC WD10EARS-00Y5B1|SLP|*|
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv4: received data doesn't have needed records
2018-05-04 07:52:13: python.d INFO: hddtemp: localipv4: check() => [FAILED]
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv6: No unix socket specified. Trying TCP/IP socket.
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv6: No request specified. Using: ""
2018-05-04 07:52:13: python.d INFO: hddtemp: localipv6: Autodetecting disks
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv6: Creating socket to "::1", port 7634
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv6: connecting socket to "::1", port 7634
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv6: Failed to connect to "::1", port 7634, error: [Errno 111] Connection refused
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv6: Traceback (most recent call last):
  File "/usr/lib/netdata/python.d/python_modules/bases/FrameworkServices/SocketService.py", line 61, in _connect2socket
    self._sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv6: closing socket
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv6: no data received
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv6: Traceback (most recent call last):
  File "/usr/lib/netdata/python.d/hddtemp.chart.py", line 77, in get_data
    raw = self._get_raw_data().split("|")[:-1]
AttributeError: 'NoneType' object has no attribute 'split'

2018-05-04 07:52:13: python.d INFO: hddtemp: localipv6: check() => [FAILED]
2018-05-04 07:52:13: python.d INFO: plugin: main: FINISHED
```

The daemon seems to be correctly running, however something could be unusual. I do not know if the following messages are the norm:
```
$ telnet localhost 7634
Trying ::1...
Connection failed: Connection refused
Trying 192.168.1.100...
Connected to localhost.
Escape character is '^]'.
|/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-HUP722020APA330_BFGWU7WF|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-WDC_WD10EARS-00Y5B1_WD-WCAV5R693922|WDC WD10EARS-00Y5B1|SLP|*|Connection closed by foreign host.
```

Running netdata 1.10.0  under ArchLinux

**Comment by l2isbad** [comment]
> |/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-HUP722020APA330_BFGWU7WF|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-WDC_WD10EARS-00Y5B1_WD-WCAV5R693922|WDC WD10EARS-00Y5B1|SLP|*|Connection closed by foreign host.

i don't see any temperature stats

**Comment by adrfantini** [comment]
That's because the disks are in sleep state (SLP). The temperature values appear correctly when they are running.

**Comment by ktsaou** [comment]
I did a test on my disks.
When the disk is in sleep or standby mode, hddtemp reports `SLP|*`.

I tested my disks when one of them is in sleep mode:

```sh
# telnet localhost 7634
Trying ::1...
Connection failed: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
|/dev/sda|Samsung SSD 840 EVO 250G B              �@|UNK|*||/dev/sdb|SanDisk SDSSDHP256G|39|C||/dev/sdc|Hitachi HDP725050GLA360|48|C||/dev/sdd|WDC WD10EZEX-08RKKA0|SLP|*|
Connection closed by foreign host.
```

(yes I know my EVO reports `UNK` with some junk characters).

sdd reports `SLP|*`.

Then checked what the plugin reports:

```sh
# /usr/libexec/netdata/plugins.d/python.d.plugin debug trace 1 hddtemp
2018-05-07 23:15:23: python.d INFO: plugin: main: Using python 3
2018-05-07 23:15:23: python.d DEBUG: plugin: main: module load source: 'hddtemp' => [OK]
2018-05-07 23:15:23: python.d DEBUG: plugin: main: module load config: 'hddtemp' => [OK]
2018-05-07 23:15:23: python.d DEBUG: plugin: main: job initialization: 'hddtemp localhost' => ['OK']
2018-05-07 23:15:23: python.d DEBUG: plugin: main: job initialization: 'hddtemp localipv4' => ['OK']
2018-05-07 23:15:23: python.d DEBUG: plugin: main: job initialization: 'hddtemp localipv6' => ['OK']
2018-05-07 23:15:23: python.d DEBUG: plugin: main: module status: 'hddtemp' => [OK] (jobs: 3)
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: No unix socket specified. Trying TCP/IP socket.
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: No request specified. Using: ""
2018-05-07 23:15:23: python.d INFO: hddtemp: localhost: Autodetecting disks
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: Creating socket to "::1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: connecting socket to "::1", port 7634
2018-05-07 23:15:23: python.d ERROR: hddtemp: localhost: Failed to connect to "::1", port 7634, error: [Errno 111] Connection refused
2018-05-07 23:15:23: python.d ERROR: hddtemp: localhost: Traceback (most recent call last):
  File "/usr/libexec/netdata/python.d/python_modules/bases/FrameworkServices/SocketService.py", line 61, in _connect2socket
    self._sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: closing socket
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: Creating socket to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: connecting socket to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: connected to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: set socket timeout to: 5.0
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: received data
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: server closed the connection
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: closing socket
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: final response: |/dev/sda|Samsung SSD 840 EVO 250G B              @|UNK|*||/dev/sdb|SanDisk SDSSDHP256G|38|C||/dev/sdc|Hitachi HDP725050GLA360|48|C||/dev/sdd|WDC WD10EZEX-08RKKA0|SLP|*|
2018-05-07 23:15:23: python.d INFO: hddtemp: localhost: check() => [OK]
2018-05-07 23:15:23: python.d INFO: hddtemp: localipv4: check() => [DROPPED] (already served by another job)
2018-05-07 23:15:23: python.d INFO: hddtemp: localipv6: check() => [DROPPED] (already served by another job)
CHART netdata.runtime_hddtemp_local '' 'Execution time for hddtemp_local' 'ms' 'python.d' netdata.pythond_runtime line 145000 1
DIMENSION run_time 'run time' absolute 1 1

2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: create() => [OK] (charts: 1)
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: started, update frequency: 1, retries: 60
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: sleeping for 0.5640078100004757 to reach frequency of 1 sec
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: Creating socket to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: connecting socket to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: connected to "127.0.0.1", port 7634
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: set socket timeout to: 5.0
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: received data
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: receiving response
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: server closed the connection
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: closing socket
2018-05-07 23:15:23: python.d DEBUG: hddtemp: localhost: final response: |/dev/sda|Samsung SSD 840 EVO 250G B              @|UNK|*||/dev/sdb|SanDisk SDSSDHP256G|38|C||/dev/sdc|Hitachi HDP725050GLA360|48|C||/dev/sdd|WDC WD10EZEX-08RKKA0|SLP|*|
CHART hddtemp_local.temperatures 'disks_temp' 'Disks Temperatures' 'Celsius' 'temperatures' 'hddtemp.temperatures' line 60000 1 '' 'python.d.plugin' 'hddtemp'
DIMENSION 'sda' 'sda' absolute 1 1 ''
DIMENSION 'sdb' 'sdb' absolute 1 1 ''
DIMENSION 'sdc' 'sdc' absolute 1 1 ''
DIMENSION 'sdd' 'sdd' absolute 1 1 ''

BEGIN hddtemp_local.temperatures 0
SET 'sda' = 0
SET 'sdb' = 38
SET 'sdc' = 48
SET 'sdd' = 0
END
```

So, it reports all disks.

The only problem I see is that `UNK` and `SLP` are reported as zero. My understanding is that they should be gaps (i.e. not reported at all until they receive a value). But this is minor.

So, auto-detection works when hddtemp reports `SLP`.

I also tested with just sdd configured in hddtemp, to make sure the plugin autodetection works when all the disks are in `SLP`. It worked again. The plugin reported the disk to netdata.

So, It is a mystery why on @adrfantini machine it reports this:

```
2018-05-04 07:52:13: python.d DEBUG: hddtemp: localipv4: final response: |/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-HUP722020APA330_BFGWU7WF|HUP722020APA330|SLP|*||/dev/disk/by-id/ata-WDC_WD10EARS-00Y5B1_WD-WCAV5R693922|WDC WD10EARS-00Y5B1|SLP|*|
2018-05-04 07:52:13: python.d ERROR: hddtemp: localipv4: received data doesn't have needed records
```

@l2isbad could you please check the source? Is it python version related?

**Comment by l2isbad** [comment]
> The only problem I see is that UNK and SLP are reported as zero.

Yes, https://github.com/firehol/netdata/blob/master/python.d/hddtemp.chart.py#L85-L88

I see
https://github.com/firehol/netdata/blob/master/python.d/hddtemp.chart.py#L47-L52

So it finds all the disks in `/dev` `startswith("sd")` and add them to `self.disks`

Then
https://github.com/firehol/netdata/blob/master/python.d/hddtemp.chart.py#L83-L84

@ktsaou and my `hddtemp` response starts with `/dev/sd`

> |/dev/sda|WDC WD10EZEX-00BN5A0|35|C|

> |/dev/sda|Samsung SSD 840 EVO 250G B

@adrfantini `hddtemp` reports

> |/dev/disk/by-id/ata-HUP722020APA330_BFJ0WS3F|HUP722020APA330|SLP|*||

**Comment by ktsaou** [comment]
hm... this explains it.

We could probably get the last strings after ` / ` from the first component as `id` and the second component as `name`. This would make it a bit more generic.

What do you think?

**Comment by ktsaou** [comment]
@l2isbad what do you think? Close this, bug or enhancement?

**Comment by l2isbad** [comment]
enhancement

**Comment by adrfantini** [comment]
Thanks, I will test, but it will take a few days.

**Comment by ilyam8** [comment]
Hi @svcadg11 

> 2019-01-17 13:16:35: python.d DEBUG: hddtemp: hddtemp: final response: |/dev/sda|DELL PERC 6/i|NA|*||/dev/sdb|DELL PERC 6/i|NA|*|
2019-01-17 13:16:35: python.d ERROR: hddtemp: hddtemp: received data doesn't have needed records

> /dev/sda|DELL PERC 6/i|NA|*||/dev/sdb|DELL PERC 6/i|NA|*|

There is nothing to chart. Am i wrong?

**Comment by svcadg11** [comment]
Hi @ilyam8,

thank you for this quick answer. Well I was still investigating about this issue, when I tried to dig deeper. I think I posted this comment to quickly :/

It seems that devices behind RAID controllers are working in a different way : 

https://serverfault.com/questions/524626/hddtemp-and-drivers

It looks like we need to use smartmontools; then we should use this : 
```
smartctl -a --device=megaraid,0 /dev/sda
```

sorry for the unconvenience.

svcadg11
